File: 0297-serialize-and-deserialize-binary-tree/0297-serialize-and-deserialize-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:
    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        if root is None:
            return "N"
        
        stack = [root]
        preorder = []
        
        ## Pre order
        while stack:
            node = stack.pop()
            if node:
                preorder.append(str(node.val))
                stack.append(node.right)  
                stack.append(node.left)
            else:
                preorder.append("N")
        return ",".join(preorder)

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        values = iter(data.split(","))
        
        def helper():
            val = next(values)
            if val == "N":
                return None
            node = TreeNode(int(val))
            node.left = helper()
            node.right = helper()
            return node
        
        return helper()
# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))

File: 0297-serialize-and-deserialize-binary-tree/0297-serialize-and-deserialize-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {

        if (root == null) {
            return "";
        }

        StringBuilder output = new StringBuilder();
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();
            
            if (node == null) {
                output.append("#,");
            } else {
                output.append(node.val).append(',');

                queue.offer(node.left);
                queue.offer(node.right);
            }
        }

        return output.toString();
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        if (data == "") {
            return null;
        }
        
        StringBuilder inputString = new StringBuilder(data);

        int firstComma = inputString.indexOf(",");
        String nodeVal = inputString.substring(0, firstComma);
        TreeNode root = new TreeNode(Integer.parseInt(nodeVal));
        inputString.delete(0, firstComma + 1);
        
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();
            firstComma = inputString.indexOf(",");

            nodeVal = inputString.substring(0, firstComma);
            inputString.delete(0, firstComma + 1);

            if (!nodeVal.equals("#")) {
                TreeNode leftNode = new TreeNode(Integer.parseInt(nodeVal));
                node.left = leftNode;
                queue.offer(leftNode);
            }            

            
            firstComma = inputString.indexOf(",");

            nodeVal = inputString.substring(0, firstComma);
            inputString.delete(0, firstComma + 1);

            if (!nodeVal.equals("#")) {
                TreeNode rightNode = new TreeNode(Integer.parseInt(nodeVal));
                node.right = rightNode;
                queue.offer(rightNode);
            }   
        }

        return root;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// TreeNode ans = deser.deserialize(ser.serialize(root));